### PR TITLE
add production approved label

### PR DESCRIPTION
**Pull request by anshupande**

### CODE REVIEW

File: aqua-ethos-setup/aqua_ethos.sh [hunk 1]
```diff
@@ -46,6 +46,14 @@ else
 	curl --silent -H "Content-Type: application/json" -H "$HEADER: Bearer $TOKEN" -X POST -d '{"name":"core-user-rule","description": "Core User is Admin of all containers","role":"administrator","resources":{"containers":["*"],"images":["*"],"volumes":["*"],"networks":["*"]},"accessors":{"users":["core"]}}' $WEB_URL/adminrules
 fi
 
+EXISTING_LABEL=$(curl --write-out %{http_code} --silent --output /dev/null -H "$HEADER: Bearer $TOKEN" $WEB_URL/settings/labels/production%20approved)
+
+if [[ "$EXISTING_LABEL" == "200" ]]; then
+	log "production approved label exists..."
+else
+	curl --silent -H "Accept: application/json" -H "Content-type: application/json" -H "$HEADER: Bearer $TOKEN" -X POST -d '{"name": "production approved", "description": "production approved images"}' $WEB_URL/settings/labels
```
Comment (iderdik) [inline aqua-ethos-setup/aqua_ethos.sh:62]: We also need to attach the policy to it tat whitelists the images

Comment (anshupande) [reply aqua-ethos-setup/aqua_ethos.sh:62]: Which API call are you referring to here?

Comment (anshupande) [reply aqua-ethos-setup/aqua_ethos.sh:62]: You mean these steps ? - Create a label (this is done)/ Attach it to image(?): https://docs.aquasec.com/v2.0/docs/labels
Creating a host batch install(?):
https://docs.aquasec.com/v2.0/docs/batch-install (there is a policy on the host for allowed labels).

Comment (iderdik) [reply aqua-ethos-setup/aqua_ethos.sh:62]: A policy must be created to:
1. Block all images from running
2. allow images that have "production approved" labels to run

FD will handle associating images to the "production approved" label. However, you need to create the policy along with the label.

File: aqua-ethos-setup/aqua_ethos.sh [hunk 2]
```diff
@@ -46,6 +46,22 @@ else
 	curl --silent -H "Content-Type: application/json" -H "$HEADER: Bearer $TOKEN" -X POST -d '{"name":"core-user-rule","description": "Core User is Admin of all containers","role":"administrator","resources":{"containers":["*"],"images":["*"],"volumes":["*"],"networks":["*"]},"accessors":{"users":["core"]}}' $WEB_URL/adminrules
 fi
 
+IMAGE_ASSURANCE=$(curl --write-out %{http_code} --silent --output /dev/null -H "$HEADER: Bearer $TOKEN" $WEB_URL/image_policy)
+
+if [[ "$IMAGE_ASSURANCE" == "200" ]]; then
+	log "image-assurance-rule exists..."
+else
+curl --silent -H "$HEADER: Bearer $TOKEN" -X POST -d '{"only_registered_images":true,"daily_scan_enabled":true,"daily_scan_time":{"hour":1,"minute":0,"second":0},"average_score_enabled":false,"average_score":7.5,"maximum_score_enabled":false,"maximum_score":8,"author":"system","cves_black_list_enabled":true,"cves_black_list":["CVE-2014-6271","CVE-2014-0160","CVE-2012-1723","CVE-2013-2465","CVE-2016-2842","CVE-2016-2108","CVE-2016-0705"],"allowed_images":[]}' $WEB_URL/image_policy
```
Comment (matthewdfuller) [inline aqua-ethos-setup/aqua_ethos.sh]: Indentation

File: aqua-ethos-setup/aqua_ethos.sh [hunk 3]
```diff
@@ -73,23 +89,25 @@ curl --silent -H "$HEADER: Bearer $TOKEN" -X PUT -d @$CRED_DIR/configs/threat_mi
 
 sudo rm -rf $CRED_DIR/configs
 
-while [[ "$EXISTING_PROFILE" == "200" && "$EXISTING_RULE" == "200" ]]; do
-	log "Profile and rule are still active..."
+while [[ "$EXISTING_PROFILE" == "200" && "EXISTING_LABEL" == "200" && "$EXISTING_RULE" == "200" ]]; do
```
Comment (matthewdfuller) [inline aqua-ethos-setup/aqua_ethos.sh]: We should add `IMAGE_ASSURANCE` here as well

**Review by matthewdfuller — CHANGES_REQUESTED** [review]

File: aqua-ethos-setup/aqua_ethos.sh (hunk 1 above)
Comment (anshupande) [reply aqua-ethos-setup/aqua_ethos.sh:62]: added 2 - allow images that have "production approved" labels to run

File: aqua-ethos-setup/aqua_ethos.sh [hunk 4]
```diff
@@ -54,6 +70,25 @@ else
 	curl --silent -H "Accept: application/json" -H "Content-type: application/json" -H "$HEADER: Bearer $TOKEN" -X POST -d '{"name": "Ethos", "type": "security.profile", "description": "Ethos Default RunTime Profile", "encrypt_all_envs": true}' $WEB_URL/securityprofiles
 fi
 
+EXISTING_ARTIFACTORY=$(curl --write-out %{http_code} --silent --output /dev/null -H "$HEADER: Bearer $TOKEN" $WEB_URL/registries/artifactory-test)
+
+if [[ "$EXISTING_ARTIFACTORY" == "200" ]]; then
+	log "Artifactory Integration exists..."
+else
+	curl --silent -H "Accept: application/json" -H "Content-type: application/json" -H "$HEADER: Bearer $TOKEN" -X POST -d  '{"prefixes": [], "auto_pull": false, "auto_pull_time": "03:00", "auto_pull_max": 100, "name": "artifactory-test", "type": "ART", "webhook": {"enabled": false, "url": "", "auth_token": ""}, "url": "$Artifactory_URL", "username": "$Artifactory_USERNAME", "password": "$Artifactory_PASSWORD"}' $WEB_URL/registries
```
Comment (matthewdfuller) [inline aqua-ethos-setup/aqua_ethos.sh]: Let's stick with capital letters for the env variables: `$ARTIFACTORY_URL`, `$ARTIFACTORY_USERNAME`, etc.

**Review by matthewdfuller — CHANGES_REQUESTED** [review]

**Comment by matthewdfuller** [comment]
Is this PR going to block our admin apps? I don't see any place where we are whitelisting those, even though they won't have the whitelist label

**Comment by anshupande** [comment]
We need to add a label `production approved` and avoid the API to whitelist the admin images as scanning them will waste lot of time. This is being followed here with aqua: https://support.aquasec.com/helpdesk/tickets/77

**Comment by msabramo** [comment]
`<Pie in the sky stuff requires="aqua-support">`
Instead of using labels (that we have to apply when we deploy a container with the image), what if we could do something when we build the image that whitelists it. For example, what if instead we put a file into the image that is cryptographically signed with an [HMAC](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) and secret key (or maybe just signed with gpg)? By putting a file into the image that is signed with our key, we are authenticating that the image is from us. Then Aqua could scan the image and look for this file and verify that it's signed properly and then maybe add the label itself.

The neat thing is that none of this needs to happen at deploy-time; it can happen at image build-time, so then it's out of the way by time we deploy.

We could maybe even borrow ideas from existing signed package systems, like Debian's for example: https://wiki.debian.org/SecureApt
`</Pie in the sky stuff>`

This of course is a big change for Aqua, but OTOH they are a security company, so they might be motivated to get package whitelisting right, because other potential customer of theirs might want it.

**Comment by msabramo** [comment]
Hmmm, just found this: https://docs.docker.com/engine/security/trust/content_trust/#image-tags-and-content-trust

> Once content trust is enabled, image publishers can sign their images. Image consumers can ensure that the images they use are signed. Publishers and consumers can be individuals alone or in organizations. Docker’s content trust supports users and automated processes such as builds.

If you set `DOCKER_CONTENT_TRUST=1`, then Docker will refuse to use images that aren't signed:

```
$ DOCKER_CONTENT_TRUST=1 docker pull behance/flight-director:latest
Error: remote trust data does not exist for docker.io/behance/flight-director: notary.docker.io does not have trust data for docker.io/behance/flight-director
```

but then I can build a signed image:

```
$ TAG=docker_content_trust_test DOCKER_CONTENT_TRUST=1 make build
...
Signing and pushing trust metadata
You are about to create a new root signing key passphrase. This passphrase
will be used to protect the most sensitive key in your signing system. Please
choose a long, complex passphrase and be careful to keep the password and the
key file itself secure and backed up. It is highly recommended that you use a
password manager to generate the passphrase and keep it safe. There will be no
way to recover this key. You can find the key in your config directory.
Enter passphrase for new root key with ID 3ecbe9c:
Passphrase is too short. Please use a password manager to generate and store a good random passphrase.
Enter passphrase for new root key with ID 3ecbe9c:
Repeat passphrase for new root key with ID 3ecbe9c:
Enter passphrase for new repository key with ID ecb8525 (docker.io/behance/flight-director):
Repeat passphrase for new repository key with ID ecb8525 (docker.io/behance/flight-director):
Finished initializing "docker.io/behance/flight-director"
Successfully signed "docker.io/behance/flight-director":c1ab4411c8582af7e19e32d78a5579a1765eb870
The push refers to a repository [docker.io/behance/flight-director]
...
docker_content_trust_test: digest: sha256:9b818630ce3907ecdfab2ad19ed346d43702518a6fa43f66cee50b314e547545 size: 1157
Signing and pushing trust metadata
Enter passphrase for repository key with ID ecb8525:
Successfully signed "docker.io/behance/flight-director":docker_content_trust_test
```

and then I can pull this signed image:

```
$ DOCKER_CONTENT_TRUST=1 docker pull behance/flight-director:docker_content_trust_test
Pull (1 of 1): behance/flight-director:docker_content_trust_test@sha256:9b818630ce3907ecdfab2ad19ed346d43702518a6fa43f66cee50b314e547545
sha256:9b818630ce3907ecdfab2ad19ed346d43702518a6fa43f66cee50b314e547545: Pulling from behance/flight-director
Digest: sha256:9b818630ce3907ecdfab2ad19ed346d43702518a6fa43f66cee50b314e547545
Status: Image is up to date for behance/flight-director@sha256:9b818630ce3907ecdfab2ad19ed346d43702518a6fa43f66cee50b314e547545
Tagging behance/flight-director@sha256:9b818630ce3907ecdfab2ad19ed346d43702518a6fa43f66cee50b314e547545 as behance/flight-director:docker_content_trust_test
```

File: aqua-ethos-setup/aqua_ethos.sh [hunk 5]
```diff
@@ -54,6 +70,25 @@ else
 	curl --silent -H "Accept: application/json" -H "Content-type: application/json" -H "$HEADER: Bearer $TOKEN" -X POST -d '{"name": "Ethos", "type": "security.profile", "description": "Ethos Default RunTime Profile", "encrypt_all_envs": true}' $WEB_URL/securityprofiles
 fi
 
+EXISTING_ARTIFACTORY=$(curl --write-out %{http_code} --silent --output /dev/null -H "$HEADER: Bearer $TOKEN" $WEB_URL/registries/artifactory-test)
+
+if [[ "$EXISTING_ARTIFACTORY" == "200" ]]; then
+	log "Artifactory Integration exists..."
+else
+	curl --silent -H "Accept: application/json" -H "Content-type: application/json" -H "$HEADER: Bearer $TOKEN" -X POST -d  '{"prefixes": [], "auto_pull": false, "auto_pull_time": "03:00", "auto_pull_max": 100, "name": "artifactory-test", "type": "ART", "webhook": {"enabled": false, "url": "", "auth_token": ""}, "url": "$ARTIFACTORY_URL", "username": "$ARTIFACTORY_USERNAME", "password": "$ARTIFACTORY_PASSWORD"}' $WEB_URL/registries
```
Comment (matthewdfuller) [inline aqua-ethos-setup/aqua_ethos.sh:78]: This will not work because you are using variables inside of single quotes in bash

**Comment by matthewdfuller** [comment]
I've pulled out the artifactory-specific work here: https://github.com/adobe-platform/dcos-scripts/pull/31

I am closing this because the file and setup has changed significantly.

Once Aqua 2.1 is tested, please submit a new PR for just the image assurance and label aspects of this PR. Do not duplicate the artifactory work from #31

**Pull request closed without merging**